### PR TITLE
Add firmware update modules

### PR DESCRIPTION
**Pull request by Mariomds**
New firmware_update module to allow the update, upgrade, and accept operations on managed systems. Also includes a new managed_systems_facts module to gather information on the HMC's managed systems. Combining the two allows the user to get multiple systems to the same target level at once.

### CODE REVIEW

File: plugins/modules/firmware_update.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+                          username=dict(required=True, type='str'),
+                          password=dict(type='str', no_log=True),
+                      )
+                      ),
+        system_name=dict(type='str', required=True),
+        action=dict(type='str', choices=['accept']),
+        state=dict(type='str', choices=['updated', 'upgraded', ]),
+        level=dict(type='str', default='latest'),
+        repository=dict(type='str', default='ibmwebsite', choices=['ibmwebsite', 'ftp', 'sftp']),
+        remote_repo=dict(type='dict', options=dict(
```
Comment (naveenkandakur) [inline plugins/modules/firmware_update.py:267]: I think this needs to be made mandatory when the state is set to updated or upgraded

File: plugins/modules/firmware_update.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                              hostname=dict(type='str'),
+                              userid=dict(type='str'),
+                              passwd=dict(type='str', no_log=True),
+                              sshkey_file=dict(type='str'),
+                              directory=dict(type='str'), )
```
Comment (naveenkandakur) [inline plugins/modules/firmware_update.py]: I think directory option should be set as required=True and same for other options also if any of them are mandatory

File: plugins/module_utils/hmc_resource.py [hunk 3]
```diff
@@ -518,3 +518,60 @@ def authenticateHMCs(self, remote_hmc, username=None, passwd=None, test=False):
...
+            if host:
+                updlic_cmd += self.OPT['UPDLIC']['-H'] + host
+            username = remote_repo['userid']
+            if username:
+                updlic_cmd += self.OPT['UPDLIC']['-U'] + username
+            passwd = remote_repo['passwd']
+            if passwd:
+                updlic_cmd += self.OPT['UPDLIC']['--PASSWD'] + passwd
+            directory = remote_repo['directory']
+            if directory:
```
Comment (naveenkandakur) [inline plugins/module_utils/hmc_resource.py]: we can remove this if condition if we set this as required=True

**Review by naveenkandakur — CHANGES_REQUESTED** [review]

File: plugins/modules/firmware_update.py (hunk 1 above)
Comment (Mariomds) [reply plugins/modules/firmware_update.py:267]: The reason it's not mandatory is because the ibmwebsite choice doesn't support the remote_repo options on updlic. So there can be update/upgrade calls that don't have any remote_repo values.

File: plugins/modules/firmware_update.py (hunk 2 above)
Comment (Mariomds) [reply plugins/modules/firmware_update.py]: I agree, updated.

File: plugins/module_utils/hmc_resource.py (hunk 3 above)
Comment (Mariomds) [reply plugins/module_utils/hmc_resource.py]: Done

File: plugins/modules/managed_systems_facts.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python
```
Comment (AnilVijayan) [inline plugins/modules/managed_systems_facts.py]: Do we really need this module? We already have `power_system `module which collects facts about individual server.
Also we do have a plan to include servers info into the inventory list using inventory plugin (customer request), pipelined for Dec release

**Review by AnilVijayan — CHANGES_REQUESTED** [review]

File: plugins/modules/managed_systems_facts.py (hunk 4 above)
Comment (Mariomds) [reply plugins/modules/managed_systems_facts.py]: The point of this new module is to gather facts from an HMC target as a whole. The case I was thinking of, for example, was when the user needed a way to automate multiple updates for a specific type of system being managed by their fleet of HMCs. This allows them to query for all managed systems (through lssysconn) in the specific inventory item (HMC) and get the entire list, which they can then filter for certain Machine Types and do all updates without having to specify all the different MTMS from a hard list that they have to filter manually. Relieving the user of that burden seems beneficial, and it also removes the need to call other HMC commands to ensure a System is acted on by the HMC that is managing it, so it's a more efficient way (speed/load wise) to handle it.

I'm not sure if the power_system could be used the same way, but I hadn't noticed it before. Do let me know if there's a way to do the above with this one, or if we can move functionality from the new module to this one. The latter didn't seem obvious because power_system seems to be used to target specific systems, so separating them made sense to keep with the Ansible doc's suggestion of "doing one thing well".
Basically, the intent is to use it for facts on the HMC about the systems being managed, while the power_system module would be used for action/facts about one system specifically.

> we do have a plan to include servers info into the inventory list using inventory plugin (customer request), pipelined for Dec release

This one might work though I'm not sure I understand how it works yet, I'll look into it. As long as the user gets a list of the managed systems of the inventory item in that iteration, and it can be used with the yaml filters to pick out a specific machine type as part of another play in the same playbook it should be the same thing (albeit probably slower due to the extra overhead of the inventory module).

Comment (AnilVijayan) [reply plugins/modules/managed_systems_facts.py]: Yes i understood the use case, and it make sense as well. But as I mentioned on earlier thread, we have a customer request to include CEC/Server info into the existing inventory plugin module (still need to evaluate the feasibility of it) and if that happens then using CEC group created by inventory plugin, user should be able to iterate and do firmware update.

Comment (Mariomds) [reply plugins/modules/managed_systems_facts.py]: > and if that happens then using CEC group created by inventory plugin, user should be able to iterate and do firmware update.

Sounds good. Module removed.
